### DISCUSSION TITLE
Running SeisSol through ParMetis. ParMetis says 'no'.

### DISCUSSION THREAD

**Discussion by LukasvdWiel** [SeisSol/SeisSol#523] [Q&A] [answered]
Dear all,

Through various discussions with developers and more experienced users I have come quite far in installing SeisSol on the cluster of the Earth Science department of Utrecht University.

I even came to the point when I can run the TPV33 test case, 
that I assume has been tested and known to run, as it came from the repo.
However, somehow, data is passed from SeisSol to ParMetis that disagrees with it, and gives a error that clarifies nothing at all:

Standard error says:
Error: mesh partitioning step failed

And standard out says:
...
Thu Mar 24 14:26:21, Info: Computing LTS weights.
Thu Mar 24 14:26:26, Info: Computing LTS weights. Done. (316 reductions.)
PARMETIS ERROR: The sum of tpwgts for constraint #0 is not 1.0

Has anybody seen this before?
Did I perhaps compile ParMetis incorrectly, or link SeisSol incorrectly to it?

**Comment by Thomas-Ulrich** [comment]
Hi, 
Probably related to this file that you need to change:
#edit ./metis/include/metis.h IDXTYPEWIDTH to be 64 (default is 32).
see:
https://seissol.readthedocs.io/en/latest/compilation.html?highlight=parmetis#installing-parmetis-optional-puml-mesh-format

**Reply by Thomas-Ulrich** [reply]
@ravil-mobile ?

**Comment by LukasvdWiel** [comment]
What happened was that I linked the proper ParMetis (and set IDXTYPEWIDTH properly)
but I failed to link the proper regular Metis with it.
It linked to a different Metis version. Once I fixed that it seems to get through the meshing phase properly!

Next step is to get the interaction with Yaml to work:

tpv33_fault.yaml: yaml-cpp: error at line 8, column 39: Unknown tag !FunctionMap
!<!EvalModel>
parameters: [Rt, radius]
model: !<!FunctionMap>
  map:
    Rt: "if (x<-9800.) {\n  Rx = (-x - 9800.0)/10000.;\n} else { \n  if (x>1100.0) {\n    Rx = (x - 1100.$
    radius: "xHypo = -6000.0;\nzHypo = -6000.0;\nreturn sqrt( ((x-xHypo)*(x-xHypo))+((z-zHypo)*(z-zHypo))$
components: !<!FunctionMap>
  map:
    s_xx: return -60000000.0;
    s_yy: return -60000000.0;
    s_zz: return  0.0;
    s_xy: "pi = 4.0 * atan (1.0);\ns_xy0 = 30000000.0*(1.0-Rt);\ns_xy1 = 0.0;\nif (radius<=550.0) {\n  s_$
    s_yz: return 0.0;
    s_xz: return 0.0;
terminate called after throwing an instance of 'std::runtime_error'
  what():  Error while parsing easi model file.
[gpu028:58651] *** Process received signal ***
[gpu028:58651] Signal: Aborted (6)
[gpu028:58651] Signal code:  (-6)

I used version 0.7.0. Is this perhaps too recent?
What version are you using?

Edit: I read there is and 'old' and a 'new' Yaml API. The old is applied in versions 0.3.x and 0.4.x and the new one in 0.5.x and higher... So if you use Yaml 0.3.x or 0.4.x to get this test case to work, we have found the answer :-)

**Reply by krenzland** [reply]
Did you compile easi as described here: https://easyinit.readthedocs.io/en/latest/getting_started.html ?
Maybe you compiled without ImpalaJIT, in this case FunctionMap is not supported. The error message could be clearer...

**Reply by LukasvdWiel** [reply]
Great hint because I indeed compiled without ImpalaJIT.
If I included ImpalaJIT, cmake could not find it because easi does not have Findimpalajit.cmake and ImpalaJIT initialy did not have impalajitConfig.cmake or impalajit-config.cmake for some reason.

Recompiling gave me a proper impalajitConfig.cmake that I could direct easi toward.

Somehow that works now. Thank you!

**Comment by LukasvdWiel** [comment, marked as answer]
Everything finally works.

Another mistake I had made was that I compiled Perl incorrectly, which was necessary to compile openSSL, which was necessary to compile Python in a way that it could install packages with PIP, which was necessary to install numpy, which was necessary to compile yateto.

After that, everything was smooth sailing.
I have a test run generating output right now. :-)

Thank you all for all the help! I could have done it without you.

**Reply by krenzland** [reply]
Great! Just out of interest, why are you doing this:

> I had made was that I compiled Perl incorrectly, which was necessary to compile openSSL, which was necessary to compile Python in a way that it could install packages with PIP, which was necessary to install numpy, which was necessary to compile yateto.

Instead of using a binary package from Python? You can also use something like anaconda which makes the installation of Numpy a lot less painful.

**Reply by LukasvdWiel** [reply]
I tried to install a Python binary directly via Anaconda, as you suggest, but it came standard with hdf5 that was too recent for SeisSol (1.12.x). I might have been able to remove it (conda remove), so that SeisSol could link to the older hdf5 of the proper version (1.8.21, following https://seissol.readthedocs.io/en/latest/compilation.html#installing-hdf5) that was manually compiled. That might have been easier indeed. It was good exercise to do this by hand, though. I should have installed Perl properly in the first step (it was an old install, when I was new to this manual procedure. I used to be in apt/yum-heaven before all this) but forgot a step there. Had I done that correctly as I would have done now, openSSL would have compiled easily, and it would all have been easier.

We run a system used by about a 170 users. All software for these users is encapsulated in modules, to prevent version conflicts. We can never upgrade something because you can rest assured that one of those users will be using some deprecated feature in that code that is not backwards compatible, and his or her research project will break down, and we have to install an older version again, which might be unavailable by the time of the upgrade. This has happened. So when a user wants a new version of anything, we just download the source of the new version, compile it, put it in a new module, create modules with its dependencies, and the user can go ahead. For example, the SeisSol module has turned to be nightmare-ish and has 12 non-generic modules on which it depends that were installed by hand, that the user had to load, for software like NUMA, TUM-Utils, Yateto, XdmfWriter, async, easi, etc. These are all in separate modules so that newer versions can be installed beside them, if the future requires it. This method is in our experience the most robust for the users.

We monitor which modules are loaded by users and when. If modules have not been loaded for a long time we can remove them to keep the environment is clean. The drawback is that it can clutter a bit: at the moment we have about 240 custom modules, but it works for everybody.

**Reply by krenzland** [reply]
Okay, that makes sense. Have you tried to compile with the newer HDF5 version? We haven't. We recommend a slightly older version because Paraview often lags behind and would then be unable to open the files.

**Reply by LukasvdWiel** [reply]
The SebastianFix515 branch that was recommended to me because it does not give the vtable error upon linking works with a newer hdf5 (1.12.1), but compiling the master with hdf5 1.12.1. gives a conflict between PUML and hdf5:

```
In file included from /trinity/opt/apps/software/SeisSol/9maart2022/src/Initializer/ParameterDB.cpp:45:
/trinity/opt/apps/software/SeisSol/9maart2022/submodules/PUML/PUML.h: In member function ‘void PUML::PUML<Topo>::open(const char*, const char*)’:
/trinity/opt/apps/software/SeisSol/9maart2022/submodules/PUML/PUML.h:185:14: error: there are no arguments to ‘H5Pset_fapl_mpio’ that depend on a template parameter, so a declaration of ‘H5Pset_fapl_mpio’ must be available [-fpermissive]
  185 |   checkH5Err(H5Pset_fapl_mpio(h5plist, m_comm, MPI_INFO_NULL));
      |              ^~~~~~~~~~~~~~~~
/trinity/opt/apps/software/SeisSol/9maart2022/submodules/PUML/PUML.h:47:37: note: in definition of macro ‘checkH5Err’
   47 | #define checkH5Err(...) _checkH5Err(__VA_ARGS__, __FILE__, __LINE__)
      |                                     ^~~~~~~~~~~
/trinity/opt/apps/software/SeisSol/9maart2022/submodules/PUML/PUML.h:185:14: note: (if you use ‘-fpermissive’, G++ will accept your code, but allowing the use of an undeclared name is deprecated)
  185 |   checkH5Err(H5Pset_fapl_mpio(h5plist, m_comm, MPI_INFO_NULL));
      |              ^~~~~~~~~~~~~~~~
/trinity/opt/apps/software/SeisSol/9maart2022/submodules/PUML/PUML.h:47:37: note: in definition of macro ‘checkH5Err’
   47 | #define checkH5Err(...) _checkH5Err(__VA_ARGS__, __FILE__, __LINE__)
      |                                     ^~~~~~~~~~~
/trinity/opt/apps/software/SeisSol/9maart2022/submodules/PUML/PUML.h:224:14: error: there are no arguments to ‘H5Pset_dxpl_mpio’ that depend on a template parameter, so a declaration of ‘H5Pset_dxpl_mpio’ must be available [-fpermissive]
  224 |   checkH5Err(H5Pset_dxpl_mpio(h5alist, H5FD_MPIO_COLLECTIVE));
      |              ^~~~~~~~~~~~~~~~
/trinity/opt/apps/software/SeisSol/9maart2022/submodules/PUML/PUML.h:47:37: note: in definition of macro ‘checkH5Err’
   47 | #define checkH5Err(...) _checkH5Err(__VA_ARGS__, __FILE__, __LINE__)
      |                                     ^~~~~~~~~~~
/trinity/opt/apps/software/SeisSol/9maart2022/submodules/PUML/PUML.h: In member function ‘void PUML::PUML<Topo>::addData(const char*, PUML::DataType)’:
/trinity/opt/apps/software/SeisSol/9maart2022/submodules/PUML/PUML.h:300:14: error: there are no arguments to ‘H5Pset_fapl_mpio’ that depend on a template parameter, so a declaration of ‘H5Pset_fapl_mpio’ must be available [-fpermissive]
  300 |   checkH5Err(H5Pset_fapl_mpio(h5plist, m_comm, MPI_INFO_NULL));
      |              ^~~~~~~~~~~~~~~~
/trinity/opt/apps/software/SeisSol/9maart2022/submodules/PUML/PUML.h:47:37: note: in definition of macro ‘checkH5Err’
   47 | #define checkH5Err(...) _checkH5Err(__VA_ARGS__, __FILE__, __LINE__)
      |                                     ^~~~~~~~~~~
/trinity/opt/apps/software/SeisSol/9maart2022/submodules/PUML/PUML.h:338:14: error: there are no arguments to ‘H5Pset_dxpl_mpio’ that depend on a template parameter, so a declaration of ‘H5Pset_dxpl_mpio’ must be available [-fpermissive]
  338 |   checkH5Err(H5Pset_dxpl_mpio(h5alist, H5FD_MPIO_COLLECTIVE));
      |              ^~~~~~~~~~~~~~~~
/trinity/opt/apps/software/SeisSol/9maart2022/submodules/PUML/PUML.h:47:37: note: in definition of macro ‘checkH5Err’
   47 | #define checkH5Err(...) _checkH5Err(__VA_ARGS__, __FILE__, __LINE__)
      |                                     ^~~~~~~~~~~
make[2]: *** [CMakeFiles/SeisSol-lib.dir/src/Initializer/ParameterDB.cpp.o] Error 1
make[1]: *** [CMakeFiles/SeisSol-lib.dir/all] Error 2
make: *** [all] Error 2

```

**Reply by krenzland** [reply]
Are you sure that you compiled HDF5 with parallel support?
https://support.hdfgroup.org/HDF5/doc/RM/RM_H5P.html#Property-SetFaplMpio

>  H5Pset_fapl_mpio is available only in the parallel HDF5 library and is not a collective function.
